crates/sui-core/src/consensus_handler.rs

```diff
@@ -646,238 +740,1639 @@ impl<C: CheckpointServiceNotify + Send + Sync> ConsensusHandler<C> {
-                            self.epoch_store
-                                .set_consensus_tx_status(position, ConsensusTxStatus::Rejected);
-                            num_rejected_user_transactions[author] += 1;
-                        }
-                        // Skip processing rejected transactions.
-                        continue;
+                if parsed.rejected {
+                    // TODO(fastpath): Add metrics for rejected transactions.
+                    if matches!(
+                        parsed.transaction.kind,
+                        ConsensusTransactionKind::UserTransaction(_)
+                    ) {
+                        self.epoch_store
+                            .set_consensus_tx_status(position, ConsensusTxStatus::Rejected);
+                        num_rejected_user_transactions[author] += 1;
                     }
+                    // Skip processing rejected transactions.
+                    // TODO(fastpath): Handle unlocking.
+                    continue;
+                }
+                if matches!(
+                    parsed.transaction.kind,
+                    ConsensusTransactionKind::UserTransaction(_)
+                ) {
+                    self.epoch_store
+                        .set_consensus_tx_status(position, ConsensusTxStatus::Finalized);
+                    num_finalized_user_transactions[author] += 1;
+                }
+                let kind = classify(&parsed.transaction);
+                self.metrics
+                    .consensus_handler_processed
+                    .with_label_values(&[kind])
+                    .inc();
+                self.metrics
+                    .consensus_handler_transaction_sizes
+                    .with_label_values(&[kind])
+                    .observe(parsed.serialized_len as f64);
+                // UserTransaction exists only when mysticeti_fastpath is enabled in protocol config.
+                if matches!(
+                    &parsed.transaction.kind,
+                    ConsensusTransactionKind::CertifiedTransaction(_)
+                        | ConsensusTransactionKind::UserTransaction(_)
+                ) {
+                    self.last_consensus_stats
+                        .stats
+                        .inc_num_user_transactions(author);
+                }
+
+                // DONE(commit-handler-rewrite): ignored external transactions must not be recorded as processed.
+                if !initial_reconfig_state.should_accept_consensus_certs() {
+                    // DONE(commit-handler-rewrite): ignore txns due to !should_accept_consensus_certs(), unless they were previously deferred
+                    // (Note: we no lnoger need to worry about the previously deferred condition, since we are only
+                    // processing newly-received transactions at this time).
                     if parsed.transaction.kind.is_user_transaction() {
```

Comment:
This line seems to be redundant to the match statement following this?
Overall I find this part of code still hard to understand.
This is partially due to that we had poor naming on ReconfigState variants and functions. Not necessarily in this PR but would be good to at least add a TODO there to fix the naming to make it clearer.